velox/exec/fuzzer/MemoryArbitrationFuzzer.cpp

```diff
@@ -0,0 +1,448 @@
+    types.push_back(vectorFuzzer_.randType(2 /*maxDepth*/));
+  }
+
+  const auto rowType = ROW(std::move(names), std::move(types));
+
+  // 1 in 10 times use empty build.
+  if (vectorFuzzer_.coinToss(0.1)) {
+    return {BaseVector::create<RowVector>(rowType, 0, pool_.get())};
+  }
+
+  // To ensure there are some matches, sample with replacement 10% of probe join
+  // keys and use these as build keys.
+  std::vector<RowVectorPtr> input;
+  for (const auto& probe : probeInput) {
+    const auto numRows = 1 + probe->size() / 10;
+    auto build = BaseVector::create<RowVector>(rowType, numRows, probe->pool());
+
+    // Pick probe side rows to copy.
+    std::vector<vector_size_t> rowNumbers(numRows);
+    for (auto i = 0; i < numRows; ++i) {
+      rowNumbers[i] = randInt(0, probe->size() - 1);
+    }
+
+    SelectivityVector rows(numRows);
+    for (auto i = 0; i < probeKeys.size(); ++i) {
+      build->childAt(i)->resize(numRows);
+      build->childAt(i)->copy(probe->childAt(i).get(), rows, rowNumbers.data());
+    }
+
+    for (auto i = 0; i < numPayload; ++i) {
+      const auto column = i + probeKeys.size();
+      build->childAt(column) =
+          vectorFuzzer_.fuzz(rowType->childAt(column), numRows);
+    }
+
+    input.push_back(build);
+  }
+
+  return input;
+}
+
+std::vector<std::string> makeNames(const std::string& prefix, size_t n) {
+  std::vector<std::string> names;
+  names.reserve(n);
+  for (auto i = 0; i < n; ++i) {
+    names.push_back(fmt::format("{}{}", prefix, i));
+  }
+  return names;
+}
+
+RowTypePtr concat(const RowTypePtr& a, const RowTypePtr& b) {
+  std::vector<std::string> names = a->names();
+  std::vector<TypePtr> types = a->children();
+
+  for (auto i = 0; i < b->size(); ++i) {
+    names.push_back(b->nameOf(i));
+    types.push_back(b->childAt(i));
+  }
+
+  return ROW(std::move(names), std::move(types));
+}
+
+core::PlanNodePtr MemoryArbitrationFuzzer::hashJoinPlan(
+    core::JoinType joinType) {
+  const auto numKeys = randInt(1, 5);
+  const std::vector<TypePtr> keyTypes = generateJoinKeyTypes(numKeys);
+  std::vector<std::string> probeKeys = makeNames("t", keyTypes.size());
+  std::vector<std::string> buildKeys = makeNames("u", keyTypes.size());
+
+  auto probeInput = generateProbeInput(probeKeys, keyTypes);
+  auto buildInput = generateBuildInput(probeInput, probeKeys, buildKeys);
+
+  auto outputColumns =
+      (core::isLeftSemiProjectJoin(joinType) ||
+       core::isLeftSemiFilterJoin(joinType) || core::isAntiJoin(joinType))
+      ? asRowType(probeInput[0]->type())->names()
+      : concat(
+            asRowType(probeInput[0]->type()), asRowType(buildInput[0]->type()))
+            ->names();
+
+  if (core::isLeftSemiProjectJoin(joinType) ||
+      core::isRightSemiProjectJoin(joinType)) {
+    outputColumns.push_back("match");
+  }
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .values(probeInput)
```

Comment:
Let's use the plan with table scan but not value nodes for better coverage? Thanks!